libical-glib/recurrence: Return arrays with the right computed size.

tintou opened the pull request:
Consumers of this API are expecting the array to already have the right size.
Do the work for them and return a well-size array.

mcclurgm commented:
I can't speak to the code, although I will say that as a consumer of the function this is what I would ideally like. It does break 1:1 compatibility with libical, though.

A few thoughts on the docs. First, since this is effectively an API change, is there a good way to indicate the version that it's changed in? Second, I think it's common form to avoid abbreviations, so "max" should be replaced with "maximum." That is a matter of opinion, though.

winterz commented on src/libical-glib/api/i-cal-recurrence.xml:
```diff
@@ -191,11 +191,18 @@
         <parameter type="ICalRecurrence *" name="recur" comment="The #ICalRecurrence"/>
         <returns type="GArray *" annotation="transfer full, element-type gshort" comment="The by_second of #ICalRecurrence."/>
-        <comment>Gets the by_second array from #ICalRecurrence. The array size if I_CAL_BY_SECOND_SIZE.</comment>
+        <comment>Gets the by_second array from #ICalRecurrence. The array has a max size of I_CAL_BY_SECOND_SIZE.</comment>
```
max -> maximum

winterz commented on src/libical-glib/api/i-cal-recurrence.xml:
```diff
@@ -238,11 +245,18 @@
         <parameter type="ICalRecurrence *" name="recur" comment="The #ICalRecurrence"/>
         <returns type="GArray *" annotation="transfer full, element-type gshort" comment="The by_minute of #ICalRecurrence."/>
-        <comment>Gets the by_minute array from #ICalRecurrence. The array size is I_CAL_BY_MINUTE_SIZE.</comment>
+        <comment>Gets the by_minute array from #ICalRecurrence. The array has a max size of I_CAL_BY_MINUTE_SIZE.</comment>
```
max -> maximum

winterz commented on src/libical-glib/api/i-cal-recurrence.xml:
```diff
@@ -285,11 +299,18 @@
         <parameter type="ICalRecurrence *" name="recur" comment="The #ICalRecurrence"/>
         <returns type="GArray *" annotation="transfer full, element-type gshort" comment="The by_hour of #ICalRecurrence."/>
-        <comment>Gets the by_hour array from #ICalRecurrence. The array size is I_CAL_BY_HOUR_SIZE.</comment>
+        <comment>Gets the by_hour array from #ICalRecurrence. The array has a max size of I_CAL_BY_HOUR_SIZE.</comment>
```
max -> maximum

winterz commented on src/libical-glib/api/i-cal-recurrence.xml:
```diff
@@ -567,11 +623,18 @@
         <parameter type="ICalRecurrence *" name="recur" comment="The #ICalRecurrence"/>
         <returns type="GArray *" annotation="transfer full, element-type gshort" comment="The by_set_pos of #ICalRecurrence."/>
-        <comment>Gets the by_set_pos array from #ICalRecurrence. The array size is I_CAL_BY_SETPOS_SIZE.</comment>
+        <comment>Gets the by_set_pos array from #ICalRecurrence. The array has a max size of I_CAL_BY_SETPOS_SIZE.</comment>
```
max -> maximum

winterz commented on src/libical-glib/api/i-cal-recurrence.xml:
```diff
@@ -520,11 +569,18 @@
         <parameter type="ICalRecurrence *" name="recur" comment="The #ICalRecurrence"/>
         <returns type="GArray *" annotation="transfer full, element-type gshort" comment="The by_month of #ICalRecurrence."/>
-        <comment>Gets the by_month array from #ICalRecurrence. The array size is I_CAL_BY_MONTH_SIZE.</comment>
+        <comment>Gets the by_month array from #ICalRecurrence. The array has a max size of I_CAL_BY_MONTH_SIZE.</comment>
```
max -> maximum

winterz commented on src/libical-glib/api/i-cal-recurrence.xml:
```diff
@@ -473,11 +515,18 @@
         <parameter type="ICalRecurrence *" name="recur" comment="The #ICalRecurrence"/>
         <returns type="GArray *" annotation="transfer full, element-type gshort" comment="The by_week_no of #ICalRecurrence."/>
-        <comment>Gets the by_week_no array from #ICalRecurrence. The array size is I_CAL_BY_WEEKNO_SIZE.</comment>
+        <comment>Gets the by_week_no array from #ICalRecurrence. The array has a max size of I_CAL_BY_WEEKNO_SIZE.</comment>
```
max -> maximum

winterz commented on src/libical-glib/api/i-cal-recurrence.xml:
```diff
@@ -426,11 +461,18 @@
         <parameter type="ICalRecurrence *" name="recur" comment="The #ICalRecurrence"/>
         <returns type="GArray *" annotation="transfer full, element-type gshort" comment="The by_year_day of #ICalRecurrence."/>
-        <comment>Gets the by_year_day array from #ICalRecurrence. The array size is I_CAL_BY_YEARDAY_SIZE.</comment>
+        <comment>Gets the by_year_day array from #ICalRecurrence. The array has a max size of I_CAL_BY_YEARDAY_SIZE.</comment>
```
max -> maximum

winterz commented:
agree with @mcclurgm to use "maximum" over "max"
and it would be nice to somehow notify people about the behavior change

winterz commented:
also, the appveyor failures on ubuntu have been fixed in the meantime.  you could rebase this patch to fix.

tintou commented:
updated with `maximum` and rebased

tintou commented:
notifying @mcrha as he is one of the biggest user of libical-glib, looking at the code of Evolution and Evolution Data Server, no change is required for now on your side.

mcrha commented:
> looking at the code of Evolution and Evolution Data Server, no change is required for now on your side.

Thanks for the notice. I briefly checked the code too, also in evolution-ews and evolution-mapi, and I did not spot any problem with the API change.

I saw in the code that an often use of the `..._get_by_..._array()` is just to count the elements, the array is allocated with no other use. Having an API to return the count of the set elements in the respective arrays could save unnecessary allocation of the array structure and its data, including filling it. I know it's out of the scope of this report.

winterz commented:
the pre-commit builds are failing.  could you rebase again please?  I think I fixed all the pre-commit issues by now.

are you planning to add the function for returning the count of set elements, as suggested by @mcrha ? or will that be handled in a separate PR?  (or not at all)?

tintou commented:
> the pre-commit builds are failing. could you rebase again please? I think I fixed all the pre-commit issues by now.
> 
> are you planning to add the function for returning the count of set elements, as suggested by @mcrha ? or will that be handled in a separate PR? (or not at all)?

Rebased. I think that Milan's suggestion can be done in a separate PR :)

winterz commented:
libical-glib-recurrence test fails

winterz commented:
@mcrha any last comment or objections?

mcrha commented:
No no, no objection from my side. The added code looks quite similar, it could deserve a private function with needed parameters, but it's only a nitpick.

Merge it as it is, please.
